src/main/java/bsh/Interpreter.java

```diff
@@ -1254,5 +1284,14 @@ public static String getSaveClassesDir() {
     public static boolean getSaveClasses()  {
         return getSaveClassesDir() != null && !getSaveClassesDir().isEmpty();
     }
-}
 
+    public void resetParser(Reader in) {
+        try {
+            this.in.close();
+        } catch (IOException x) {
+            // nothing to do...
+        }
+        this.in = in;
+        parser = new Parser(in);
+    }
+}
```

Comment:
This should not be necessary, there is no need to reset the parser. Interrupt should interrupt the running thread not just the parser.

Reply:
This is where we have different views. I really think we should not stop REPL and recreate everything. I have seen your other comment about closing and recreating the parser not costing to much, but it simply does not sound correct. It may be a workaround to achieve our goal, but I believe there is a better way. Interpreter shall have the responsibility to get the input parse it and run the parsed code keeping the state of all this until you have finished your session.
The Parser object is what is used to parse a portion of the input and its lifecycle is shorter than Interpreter. We want to stop this last part, not everything.

said that, I was re-reading something about Thread.interrupt, and we may have the solution that works for both: 
>>
The Interrupt Status Flag

The interrupt mechanism is implemented using an internal flag known as the interrupt status. Invoking Thread.interrupt sets this flag. When a thread checks for an interrupt by invoking the static method Thread.interrupted, interrupt status is cleared. The non-static isInterrupted method, which is used by one thread to query the interrupt status of another, does not change the interrupt status flag.
<<
If we use Thread.interrupt to tell everything to stop what they are doing and then we clear the status without exiting from REPL, we would be good. The only caveat will still be we need to create a new parser, which to me seems perfectly fine.